### PR TITLE
AArch64: Implementation of BNDCHK Evaluator

### PR DESCRIPTION
**Pull request by marufunb**
This commit implements BNDCHK Evaluator for openj9.

Signed-off-by: marufunb <mrahma15@unb.ca>

### CODE REVIEW

**Comment by 0xdaryl** [comment]
This PR is pretty much a duplication of the code on x86, some of which won't build or function on AArch64.  My apologies for being blunt, but I'd rather not spend time reviewing this.

What is required is a much simpler implementation targeted to AArch64.  To begin with, you need to perform an unsigned compare of the two children of the BNDCHK (the index and the arraylength of the array object).  If the index is above or equal to the arraylength, then branch to a snippet that throws an AIOB exception (the same helper call snippet that DIVCHK and NULLCHK should use).  The one thing to be aware of is that sometimes the optimizer will constant propagate either the index or the arraylength based on its analyses.  In such cases you need to handle the possibility of a constant (immediate) operand to the instruction, possibly swapping the operands (and therefore the condition test) to fit the immediate instruction form.

**Comment by marufunb** [comment]
> This PR is pretty much a duplication of the code on x86, some of which won't build or function on AArch64. My apologies for being blunt, but I'd rather not spend time reviewing this.
> 
> What is required is a much simpler implementation targeted to AArch64. To begin with, you need to perform an unsigned compare of the two children of the BNDCHK (the index and the arraylength of the array object). If the index is above or equal to the arraylength, then branch to a snippet that throws an AIOB exception (the same helper call snippet that DIVCHK and NULLCHK should use). The one thing to be aware of is that sometimes the optimizer will constant propagate either the index or the arraylength based on its analyses. In such cases you need to handle the possibility of a constant (immediate) operand to the instruction, possibly swapping the operands (and therefore the condition test) to fit the immediate instruction form.

I have not yet completed for review Daryl. I am still working on it that's why I did not tag you or informed you. I read your message on slack and still working on this. I am also working on the build of this PR. I will let you know when it is ready for your review.

File: runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -249,3 +249,68 @@ J9::ARM64::TreeEvaluator::arraylengthEvaluator(TR::Node *node, TR::CodeGenerator
...
+      }
+   else
+      {
+      src1Reg = cg->evaluate(firstChild);
+
+      if ((secondChild->getOpCode().isLoadConst())
+            && (NULL == secondChild->getRegister()))
+         {
+         value = secondChild->getInt();
+         if (0 > value || 0x80000000 < value)
```
Comment (0xdaryl) [inline runtime/compiler/aarch64/codegen/J9TreeEvaluator.cpp]: I think you want to evaluate the child into a register only if the value is outside the 12-bit immediate range of the instruction.  This test is backwards.  See `constantIsUnsignedImm12`.

**Comment by marufunb** [comment]
@0xdaryl I have made those changes according to your feedback.

**Comment by 0xdaryl** [comment]
Thanks.  Can you confirm it builds successfully with OpenJ9 ?

**Comment by marufunb** [comment]
@0xdaryl I built it on Arch Linux (following Aaron's guidelines) for Openj9.

**Comment by 0xdaryl** [comment]
This is an AArch64-specific change.  No Jenkins required.

**Comment by knn-k** [comment]
See #6903 for a fix to this evaluator.

**Pull request merged**